src/aten/sycl/BatchNormKernels.cpp

```diff
@@ -0,0 +1,3965 @@
+      return SIMD16;
+    }
+  }
+  return simd;
+}
+
+template <typename scalar_t, typename accscalar_t>
+struct Float2 {
+  accscalar_t v1, v2;
+  Float2() {}
+
+  Float2(scalar_t v1, scalar_t v2)
+      : v1(static_cast<accscalar_t>(v1)), v2(static_cast<accscalar_t>(v2)) {}
+  Float2(int v)
+      : v1(static_cast<accscalar_t>(v)), v2(static_cast<accscalar_t>(v)) {}
+  Float2& operator+=(const Float2& a) {
+    v1 += a.v1;
+    v2 += a.v2;
+    return *this;
+  }
+
+  friend Float2 operator+(Float2 a, const Float2& b) {
+    a += b;
+    return a;
+  }
+};
+
+template <typename scalar_t, typename accscalar_t, typename PTA>
+struct GradOp {
+  GradOp(accscalar_t m, const PTA& i, const PTA& g)
+      : mean(m), input(i), grad_output(g) {}
+  Float2<scalar_t, accscalar_t> operator()(int batch, int plane, int n) const {
+    accscalar_t g = grad_output[batch][plane][n];
+    accscalar_t c = static_cast<accscalar_t>(input[batch][plane][n]) - mean;
+    return Float2<scalar_t, accscalar_t>(g, g * c);
+  }
+  const accscalar_t mean;
+  const PTA& input;
+  const PTA& grad_output;
+};
+
+template <
+    int SIMD,
+    typename accscalar_t,
+    typename reduce_op,
+    typename item_t,
+    typename local_shared_t>
+static inline void group_reduce(
+    item_t item,
+    int sub_group_num,
+    accscalar_t& val,
+    accscalar_t init,
+    const local_shared_t& local_data,
+    reduce_op bin_op) {
+  auto sg = item.get_sub_group();
+  uint32_t lane_id = sg.get_local_linear_id();
+  uint32_t sg_id = sg.get_group_linear_id();
+
+  // dynamic get SIMD width result in big performance drop
+  // uint32_t SIMD = sg.get_local_range()[0];
+#pragma unroll
+  for (int i = 1; i < SIMD; i <<= 1) {
+    val = bin_op(val, static_cast<accscalar_t>(sg.shuffle_down(val, i)));
+  }
+  if (sub_group_num == 1) {
+    if (lane_id == 0) {
+      local_data[0] = val;
+    }
+    item.barrier(sycl_local_fence);
+    val = local_data[0];
+
+    return;
+  }
+
+  // reduce internal each subgroup, each subgroup will generate one result
+  // there are WGroupSize/subGroupSize elements after this step
+  if (lane_id == 0) {
+    local_data[sg_id] = val;
+  }
+  item.barrier(sycl_local_fence);
+
+  // use one subgroup to reduce WGroupSize/subGroupSize elements
+  // into the final result
+  if (sg_id == 0) {
+    val = init;
+    if (lane_id < sub_group_num) {
+      val = accscalar_t(local_data[lane_id]);
+    }
+    for (int i = lane_id + SIMD; i < sub_group_num; i += SIMD) {
+      val = bin_op(val, static_cast<accscalar_t>(local_data[i]));
+    }
+#pragma unroll
+    for (int i = 1; i < SIMD; i <<= 1) {
+      val = bin_op(val, static_cast<accscalar_t>(sg.shuffle_down(val, i)));
+      if (i >= ((sub_group_num + 1) >> 1))
+        break;
+    }
+
+    // the 0th WI (the 0th WI in the 0th sub_group) generate the final
+    // result
+    if (lane_id == 0) {
+      local_data[0] = val;
+    }
+  }
+
+  item.barrier(sycl_local_fence);
+  val = local_data[0];
```

Comment:
Codes fail when num_of_sg > sg_size. A sg could not handle all local results yielded by sgs in the last step.

Reply:
Please,
1. Change the name of the util, to make the semantics clear. Like `group_reduce_with_limited_num_of_sg`. The name might be better.
2. Or make the util general to avoid misunderstanding when some one use it in future.

Reply:
The `get_max_group_size` util in batch_norm will obtain a constrained work group size to ensure that the amount of data processed can be covered by two subgroup reduces.

Reply:
Isee, so can you change the name or add comments for the limitation of the usage?